pkg/virt-api/rest/subresource_test.go

```diff
@@ -475,6 +476,333 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 		})
 	})
 
+	Context("Subresource api - error handling for RestartVMRequestHandler", func() {
+		BeforeEach(func() {
+			request.PathParameters()["name"] = "testvm"
+			request.PathParameters()["namespace"] = "default"
+
+			vmi = v1.VirtualMachineInstance{
+				Spec: v1.VirtualMachineInstanceSpec{},
+			}
+			vmi.ObjectMeta.SetUID(uuid.NewUUID())
+		})
+
+		It("should fail on VM with RunStrategyHalted", func(done Done) {
+			vm := newVirtualMachineWithRunStrategy(v1.RunStrategyHalted)
+
+			server.AppendHandlers(
+				ghttp.CombineHandlers(
+					ghttp.VerifyRequest("GET", "/apis/kubevirt.io/v1alpha3/namespaces/default/virtualmachines/testvm"),
+					ghttp.RespondWithJSONEncoded(http.StatusOK, vm),
+				),
+			)
+
+			app.RestartVMRequestHandler(request, response)
+
+			Expect(response.Error()).To(HaveOccurred())
+			Expect(response.StatusCode()).To(Equal(http.StatusBadRequest))
+			close(done)
+		})
+
+		It("should not fail with VMI and other RunStrategies", func(done Done) {
+			runStrategies := []v1.VirtualMachineRunStrategy{
+				v1.RunStrategyAlways,
+				v1.RunStrategyManual,
+				v1.RunStrategyRerunOnFailure,
+			}
+
+			for _, runStrategy := range runStrategies {
+				vm := newVirtualMachineWithRunStrategy(runStrategy)
+
+				server.AppendHandlers(
+					ghttp.CombineHandlers(
+						ghttp.VerifyRequest("GET", "/apis/kubevirt.io/v1alpha3/namespaces/default/virtualmachines/testvm"),
+						ghttp.RespondWithJSONEncoded(http.StatusOK, vm),
+					),
+				)
+
+				server.AppendHandlers(
+					ghttp.CombineHandlers(
+						ghttp.VerifyRequest("GET", "/apis/kubevirt.io/v1alpha3/namespaces/default/virtualmachineinstances/testvm"),
+						ghttp.RespondWithJSONEncoded(http.StatusOK, vmi),
+					),
+				)
+
+				server.AppendHandlers(
+					ghttp.CombineHandlers(
+						ghttp.VerifyRequest("PATCH", "/apis/kubevirt.io/v1alpha3/namespaces/default/virtualmachines/testvm"),
+						ghttp.RespondWithJSONEncoded(http.StatusOK, vm),
+					),
+				)
+
+				app.RestartVMRequestHandler(request, response)
+
+				Expect(response.Error()).NotTo(HaveOccurred())
+				Expect(response.StatusCode()).To(Equal(http.StatusAccepted))
+			}
+			close(done)
+		})
+
+		It("should fail anytime without VMI", func(done Done) {
+			runStrategies := []v1.VirtualMachineRunStrategy{
```

Comment:
note that ginkgo supports table driven tests: https://onsi.github.io/ginkgo/#table-driven-tests. That makes it easier to grasp if scenarios are tested or not.

Reply:
@rmohr have just seen kubevirt#1048, so I plan to remove the async testing in `subresource_test.go` also? Does that make sense for you?

Reply:
Definitely! :+1: